integration-libs/opf/quick-buy/core/services/opf-quick-buy-transaction.service.ts

```diff
@@ -0,0 +1,177 @@
+import {
+  CheckoutBillingAddressFacade,
+  CheckoutDeliveryAddressFacade,
+  CheckoutDeliveryModesFacade,
+} from '@spartacus/checkout/base/root';
+import {
+  Address,
+  BaseSiteService,
+  QueryState,
+  RoutingService,
+  UserAddressService,
+} from '@spartacus/core';
+import { OpfGlobalMessageService } from '@spartacus/opf/base/root';
+import {
+  OPF_QUICK_BUY_DEFAULT_MERCHANT_NAME,
+  OpfQuickBuyDeliveryInfo,
+  OpfQuickBuyDeliveryType,
+  OpfQuickBuyLocation,
+} from '@spartacus/opf/quick-buy/root';
+import { Observable } from 'rxjs';
+import { filter, map, switchMap, take } from 'rxjs/operators';
+@Injectable({
+  providedIn: 'root',
+})
+export class OpfQuickBuyTransactionService {
+  protected baseSiteService = inject(BaseSiteService);
+  protected activeCartFacade = inject(ActiveCartFacade);
+  protected routingService = inject(RoutingService);
+  protected checkoutDeliveryModesFacade = inject(CheckoutDeliveryModesFacade);
+  protected checkoutDeliveryAddressFacade = inject(
+    CheckoutDeliveryAddressFacade
+  );
+  protected checkoutBillingAddressFacade = inject(CheckoutBillingAddressFacade);
+  protected userAddressService = inject(UserAddressService);
+  protected opfGlobalMessageService = inject(OpfGlobalMessageService);
+
+  getTransactionDeliveryType(): Observable<OpfQuickBuyDeliveryType> {
+    return this.activeCartFacade.hasDeliveryItems().pipe(
+      take(1),
+      map((hasDeliveryItems: boolean) =>
+        hasDeliveryItems
+          ? OpfQuickBuyDeliveryType.SHIPPING
+          : OpfQuickBuyDeliveryType.PICKUP
+      )
+    );
+  }
+
+  getTransactionDeliveryInfo(): Observable<OpfQuickBuyDeliveryInfo> {
+    const deliveryTypeObservable = this.getTransactionDeliveryType().pipe(
+      map((deliveryType) => {
+        return {
+          type: deliveryType,
+        } as OpfQuickBuyDeliveryInfo;
+      })
+    );
+
+    return deliveryTypeObservable.pipe(take(1));
+  }
+
+  getTransactionLocationContext(): Observable<OpfQuickBuyLocation> {
+    return this.routingService.getRouterState().pipe(
+      take(1),
+      map(
+        (routerState) =>
+          routerState?.state?.semanticRoute?.toLocaleUpperCase() as OpfQuickBuyLocation
+      )
+    );
+  }
+
+  getMerchantName(): Observable<string> {
+    return this.baseSiteService.get().pipe(
+      take(1),
+      map((baseSite) => baseSite?.name ?? OPF_QUICK_BUY_DEFAULT_MERCHANT_NAME)
+    );
+  }
+
+  checkStableCart(): Observable<boolean> {
+    return this.activeCartFacade.isStable().pipe(
+      filter((isStable) => !!isStable),
+      take(1)
+    );
+  }
+
+  getSupportedDeliveryModes(): Observable<DeliveryMode[]> {
+    return this.checkoutDeliveryModesFacade.getSupportedDeliveryModes();
+  }
+
+  setDeliveryAddress(address: Address): Observable<string> {
+    this.opfGlobalMessageService.disableGlobalMessage([
+      'addressForm.userAddressAddSuccess',
+    ]);
+    return this.checkoutDeliveryAddressFacade.createAndSetAddress(address).pipe(
```

Comment:
as an improvement we use createAndSetAddress if no addr yet, then use PUT addresses/deliver to update it, it would avoid the temp and stale addresses.
I remember trying that last year, but if occ was improved, we could give a try.

Reply:
We can spike this in another ticket. Current one was rather to revert a previous solution not dig deeper into refactoring direction. Thanks for the suggestion and I will create a separate ticket for this.